stix2/registration.py

```diff
@@ -3,10 +3,83 @@
 from . import registry
 from .base import _DomainObject, _Observable
 from .exceptions import DuplicateRegistrationError
-from .properties import _validate_type
-from .utils import PREFIX_21_REGEX, get_class_hierarchy_names
+from .properties import (
+    ListProperty, ObjectReferenceProperty, ReferenceProperty, _validate_type,
+)
+from .utils import PREFIX_21_REGEX
 from .version import DEFAULT_VERSION
 
+# Properties ending in "_ref/s" need to be instances of specific types to meet
+# our interpretation of section 3.1 of the spec
+_VERSION_REF_TYPES_MAP = {
+    # TODO: The 2.0 code seems to use `ReferenceProperty`s - is that okay?
+    "2.0": (ObjectReferenceProperty, ReferenceProperty),
+    "2.1": (ReferenceProperty,),
+}
+# Any unknown versions are presumed to be newer than STIX 2.1 so we'll define
+# this to enforce the 2.1 compatible type in such situations
+_UNKNOWN_VERSION_REF_TYPES = _VERSION_REF_TYPES_MAP["2.1"]
+
+
+def _validate_ref_props(props_map, version):
+    """
+    Validate that reference properties contain an expected type.
+
+    Args:
+        props_map (mapping): A mapping of STIX object properties to be checked.
+        version (str): Which STIX2 version the properties must confirm to.
+
+    Raises:
+        ValueError: If the properties do not conform.
+    """
+    try:
+        ref_prop_types = _VERSION_REF_TYPES_MAP[version]
+    except KeyError:
+        ref_prop_types = _UNKNOWN_VERSION_REF_TYPES
+
+    for prop_name, prop_obj in props_map.items():
+        tail = prop_name.rsplit("_", 1)[-1]
+        if tail == "ref" and not isinstance(prop_obj, ref_prop_types):
```

Comment:
Alternatively, if we turned `ReferenceProperty` (or all `Property`/`STIXBase` types?) use the `abc.ABCMeta` metaclass, it would be possible to do `ReferenceProperty.register(MyCustomReferenceType)` and have this subclass check pass for registered virtual subclasses.

That approach would introduce a new method on those STIX object types which could arguably collide with some STIX (custom or future) attribute, which probably isn't great. Alternatively we could implement a `ReferenceProperty.__subclasshook__()` where we codify what a property needs to look like to be considered valid for use as a property ending in `_ref(s)`. However, that would result in some non-obvious data validation hidden behind these `isinstance()` calls by being implemented in the `__subclasshook__()`. Not sure I particularly like it because of that.

Reply:
As a note for myself, `stix2.properties.Property` is not a subclass of `_STIXBase`

Reply:
I think you're suggesting having different ReferenceProperty subclasses for enforcing different ref property requirements?  I don't think the requirements on ref properties are complicated enough that we need specialized custom code for different cases.  The black/whitelist on STIX types and type classes seems sufficient.  Can you think of a case where that is not sufficient?

A stylistic note: I don't like using exception handling for non-exceptional circumstances, or if the same thing could be easily done in a different way.  Instead of the KeyError exception stuff, how about:

```python
ref_prop_types = _VERSION_REF_TYPES_MAP.get(version, _UNKNOWN_VERSION_REF_TYPES)
```

?

Reply:
> I think you're suggesting having different ReferenceProperty subclasses for enforcing different ref property requirements? I don't think the requirements on ref properties are complicated enough that we need specialized custom code for different cases. The black/whitelist on STIX types and type classes seems sufficient. Can you think of a case where that is not sufficient?
> I think you're suggesting having different ReferenceProperty subclasses for enforcing different ref property requirements? I don't think the requirements on ref properties are complicated enough that we need specialized custom code for different cases. The black/whitelist on STIX types and type classes seems sufficient. Can you think of a case where that is not sufficient?

The background in #511 which this change is in support of may be illustrative, in particular this https://github.com/oasis-open/cti-python-stix2/issues/511#issuecomment-868015761 . The tl;dr is that I have a thing which I consider a reference property (reference plus some "value-add") but which is a subclass of `DictionaryProperty` rather that `ReferenceProperty`. I could do some trickery with multiple inheritance and avoiding the use of `super()` but I really don't want to do that and would prefer to be able to register my `AnnotatedReference` as a virtual subclass using the ABCMeta machinery. This would be a change which is probably exclusively useful for supporting my exotic use case though, which is why I've not implemented that in this MR without us landing on a mutually acceptable approach :)

> Instead of the KeyError exception stuff, how about ...
Sure, will do.